src/k8s/pkg/k8sd/features/cilium/network.go

```diff
@@ -201,3 +212,43 @@ func rolloutRestartCilium(ctx context.Context, snap snap.Snap, attempts int) err
 
 	return nil
 }
+
+// Ensures that the provided path is a directory with the appropriate
+// ownership/permissions for it to be used as the CNI binary directory.
+// https://github.com/canonical/k8s-snap/issues/567
+// https://github.com/cilium/cilium/issues/23838
+func ensureCniBinDir(ctx context.Context, cniBinDir string) error {
+	var stat syscall.Stat_t
+	if err := syscall.Stat(cniBinDir, &stat); err != nil {
+		if errors.Is(err, fs.ErrNotExist) {
+			log.FromContext(ctx).WithValues("cniBinDir", cniBinDir).Info("cniBinDir doesn't exist, creating now")
+			if err := os.MkdirAll(cniBinDir, 0o0700); err != nil {
+				return fmt.Errorf("failed to os.MkdirAll(%s): %w", cniBinDir, err)
+			}
+
+			if err := syscall.Stat(cniBinDir, &stat); err != nil {
+				return fmt.Errorf("failed to syscall.Stat(%q): %w", cniBinDir, err)
+			}
+		}
+		return fmt.Errorf("failed to syscall.Stat(%q): %w", cniBinDir, err)
+	}
+
+	if stat.Uid != 0 || stat.Gid != 0 {
+		msg := fmt.Sprintf("changing cniBinDir ownership from %d:%d to 0:0", stat.Uid, stat.Gid)
+		log.FromContext(ctx).WithValues("cniBinDir", cniBinDir).Info(msg)
```

Comment:
prefer a static log message, then add arguments as values (e.g. "from": "100:100" "to":"0:0")